Contributing guide & code of conduct (#692)
huggingface/huggingface_hub: CONTRIBUTING.md

on the awesome projects it made possible, shout out on Twitter every time it has
helped you, or simply star the repo to say "thank you".

Whichever way you choose to contribute, please be mindful to respect our
[code of conduct](https://github.com/huggingface/huggingface_hub/blob/master/CODE_OF_CONDUCT.md).

## You can contribute in so many ways!

The library is split in different parts, where we welcome contributions.

### The client library, `huggingface_hub`

This repository hosts the client library `huggingface_hub` which serves as a frontend to the Hugging Face Hub.
this part lives in `src/huggingface_hub`.

There are many ways you can contribute to this client library:
* Fixing outstanding issues with the existing code;
* Contributing to the examples or to the documentation;
* Submitting issues related to bugs or desired new features.

When opening a PR on this part of the repository, please add as reviewer:
- @LysandreJik
- @osanseviero

Additionally, here are the owners of specific parts of the library:
- Core of the library: @muellerzr
- Documentation: @stevhliu
- Mixins: @nateraw

### The community inference API

> In the `api-inference-community` lives a tool to enable third-party libraries support integrated with the Hugging

Reviewer: ```suggestion
The `api-inference-community` folder contains a tool to enable third-party library support integrated with the Hugging
```

> In the `api-inference-community` lives a tool to enable third-party libraries support integrated with the Hugging
> Face Hub.

Reviewer 2: ```suggestion
In the `api-inference-community` lives a tool to enable [third-party libraries](https://huggingface.co/docs/hub/libraries) support in the Inference API of the Hugging Face Hub, enabling widgets in model repositories and HTTP requests.
```

> We welcome contributions to add new containers for new libraries, to update the existing ones, and to provide help

Reviewer 2: ```suggestion
We welcome contributions to [add new containers](https://huggingface.co/docs/hub/adding-a-library#set-up-the-inference-api) for new libraries, to update the existing ones, and to provide help
```

fixing bugs and adding features.

When opening a PR on this part of the repository, please add @Narsil as a reviewer.

## JavaScript content

> In the `js` folder live the JavaScript code required to run the inference API widgets on the Hugging Face Hub.

Reviewer 2: We actually have more things than the widgets open sourced.

```suggestion
In the `js` folder live the JavaScript code of the Hub. It includes:
* The widgets ([code](https://github.com/huggingface/huggingface_hub/tree/main/js/src/lib/components/InferenceWidget)
* Code snippets to make inference calls ([code](https://github.com/huggingface/huggingface_hub/tree/main/js/src/lib/inferenceSnippets)
* Code snippets to load models ([code](https://github.com/huggingface/huggingface_hub/blob/main/js/src/lib/interfaces/Libraries.ts))
```

Here too, we welcome any logic and documentation contributions.

When opening a PR on this part of the repository, please add @mishig25 and @osanseviero as a reviewer.

*All contributions are equally valuable to the community.*

## Submitting a new issue or feature request

Do your best to follow these guidelines when submitting an issue or a feature
request. It will make it easier for us to come back to you quickly and with good
feedback.

### Did you find a bug?

The 🤗 Hugging Face Hub library is robust and reliable thanks to the users who notify us of
the problems they encounter. So thank you for reporting an issue.

First, we would really appreciate it if you could **make sure the bug was not
already reported** (use the search bar on Github under Issues).

Did not find it? :( So we can act quickly on it, please follow these steps:

* Include your **OS type and version**, the versions of **Python**, **PyTorch** and
  **Tensorflow** when applicable;
* A short, self-contained, code snippet that allows us to reproduce the bug in
  less than 30s;
* Provide the *full* traceback if an exception is raised.

> ### Do you want a new feature (that is not a model)?

Reviewer 2: ```suggestion
### Do you want a new feature?
```

A world-class feature request addresses the following points:

1. Motivation first:
* Is it related to a problem/frustration with the library? If so, please explain

>   why. Providing a code snippet that demonstrates the problem is best.

Reviewer: ```suggestion
  why and provide a code snippet that demonstrates the problem best.
```

   $ git remote add upstream https://github.com/huggingface/huggingface_hub.git
   ```

3. Create a new branch to hold your development changes:

   ```bash
   $ git checkout -b a-descriptive-name-for-my-changes
   ```

   **Do not** work on the `master` branch.

4. Set up a development environment by running the following command in a virtual environment:

   ```bash
   $ pip install -e ".[dev]"
   ```

   (If huggingface_hub was already installed in the virtual environment, remove
   it with `pip uninstall huggingface_hub` before reinstalling it in editable
   mode with the `-e` flag.)

5. Develop the features on your branch.

   As you work on the features, you should make sure that the test suite
   passes. You should run the tests impacted by your changes like this (see 
   below an explanation regarding the environment variable):

   ```bash
   $ HUGGINGFACE_CO_STAGING=1 pytest tests/<TEST_TO_RUN>.py
   ```

   You can also run the full suite with the following command.

   ```bash
   $ make test
   ```

   `hugginface_hub` relies on `black` and `isort` to format its source code
   consistently. After you make changes, apply automatic style corrections and code verifications

>    that can't be automated in one go with:

Reviewer: I think we forgot this? :)

```bash
$ make fixup
```

Author: This one is especially useful for repositories which have a lot of files. It is not the case for `huggingface_hub`, so it is not in the `Makefile`: https://github.com/huggingface/huggingface_hub/blob/main/Makefile

Reviewer: Ah ok, I thought there was supposed to be a command after the colon. Maybe it'd be clearer to say:

`After you make changes, apply automatic style corrections and code verifications that can't be automated.`

Author: Ah, I see what you mean! Great catch indeed, thank you.

   consulting the issue know you are working on it);
3. To indicate a work in progress please prefix the title with `[WIP]`. These
   are useful to avoid duplicated work, and to differentiate it from PRs ready
   to be merged;
4. Make sure existing tests pass;
5. Add high-coverage tests. No quality testing = no merge.
7. Due to the rapidly growing repository, it is important to make sure that no files that would significantly weigh down the repository are added. This includes images, videos and other non-text files. We prefer to leverage a hf.co hosted `dataset` like
   the ones hosted on [`hf-internal-testing`](https://huggingface.co/hf-internal-testing) in which to place these files and reference
   them by URL. We recommend putting them in the following dataset: [huggingface/documentation-images](https://huggingface.co/datasets/huggingface/documentation-images).
   If an external contribution, feel free to add the images to your PR and ask a Hugging Face member to migrate your images
   to this dataset.

### Tests

An extensive test suite is included to test the library behavior and several examples. Library tests can be found in

> the [tests folder](https://github.com/huggingface/huggingface_hub/tree/master/tests).

Reviewer 2: ```suggestion
the [tests folder](https://github.com/huggingface/huggingface_hub/tree/main/tests).
```

<!---
Copyright 2020 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

> -->

Author: There is also a LICENSE at the root of the repo. We have an additional license at the top of every file, with the appropriate copyrights for the code in that specific file. If some code was taken from somewhere else or from another org, we add a mention to that org in the copyright here.

It is probably less relevant here than it is for `transformers`, in which a lot of code is inspired from other repositories.

others, reaching out and improving the documentations are immensely valuable to
the community.

It also helps us if you spread the word: reference the library from blog posts
on the awesome projects it made possible, shout out on Twitter every time it has
helped you, or simply star the repo to say "thank you".

Whichever way you choose to contribute, please be mindful to respect our
[code of conduct](https://github.com/huggingface/huggingface_hub/blob/master/CODE_OF_CONDUCT.md).

## You can contribute in so many ways!

The repository is split into different parts, where we welcome contributions.

### The client library, `huggingface_hub`

This repository hosts the client library `huggingface_hub`, which is a frontend to the Hugging Face Hub.
this part lives in `src/huggingface_hub` and `tests`.

There are many ways you can contribute to this client library:
* Fixing outstanding issues with the existing code;
* Contributing to the examples or to the documentation;
* Submitting issues related to bugs or desired new features.

When opening a PR on this part of the repository, please add as reviewer:
- @LysandreJik

> - @osanseviero

Reviewer 3: What do we think of automatically adding them as a reviewer, by listing you both in a CODEOWNERS file? 

We can then also even do the specific owners of parts of the library, so if it was touched automatically myself or Steven are put on as reviewers. No need for them to have to remember to do this.

> - @osanseviero

Reviewer 3: Addressed with: https://github.com/huggingface/huggingface_hub/pull/723

@LysandreJik I think this part can be completely taken out now with this as well, since it's automatically done.

* Code snippets to make inference calls ([code](https://github.com/huggingface/huggingface_hub/tree/main/js/src/lib/inferenceSnippets)
* Code snippets to load models ([code](https://github.com/huggingface/huggingface_hub/blob/main/js/src/lib/interfaces/Libraries.ts))

Here too, we welcome any logic and documentation contributions.

When opening a PR on this part of the repository, please add @mishig25 and @osanseviero as a reviewer.

*All contributions are equally valuable to the community.*

## Submitting a new issue or feature request

Do your best to follow these guidelines when submitting an issue or a feature
request. It will make it easier for us to come back to you quickly and with good
feedback.

### Did you find a bug?

The 🤗 Hugging Face Hub library is robust and reliable thanks to the users who notify us of
the problems they encounter. So thank you for reporting an issue.

First, we would really appreciate it if you could **make sure the bug was not
already reported** (use the search bar on Github under Issues).

Did not find it? :( So we can act quickly on it, please follow these steps:

> * Include your **OS type and version**, the versions of **Python**, **PyTorch** and

Reviewer 3: Perhaps we could also include a snippet either via a template for a Bug Report, or just including the snippet here of how they can go and get those quickly.

  **Tensorflow** when applicable;
* A short, self-contained, code snippet that allows us to reproduce the bug in
  less than 30s;

> * Provide the *full* traceback if an exception is raised.

Reviewer 3: Perhaps change the wording to be include a copy (as text) of the *full* traceback....

So they don't just add screenshots of it (this I've seen more beginners do)

   control runs in CI, however you can also run the same checks with:

   ```bash
   $ make quality
   ```

   Once you're happy with your changes, add changed files using `git add` and
   make a commit with `git commit` to record your changes locally:

   ```bash
   $ git add modified_file.py
   $ git commit
   ```

   Please write [good commit messages](https://chris.beams.io/posts/git-commit/).

   It is a good idea to sync your copy of the code with the original
   repository regularly. This way you can quickly account for changes:

   ```bash
   $ git fetch upstream
   $ git rebase upstream/master
   ```

   Push the changes to your account using:

   ```bash
   $ git push -u origin a-descriptive-name-for-my-changes
   ```

6. Once you are satisfied (**and the checklist below is happy too**), go to the
   webpage of your fork on GitHub. Click on 'Pull request' to send your changes
   to the project maintainers for review.

7. It's ok if maintainers ask you for changes. It happens to core contributors
   too! So everyone can see the changes in the Pull request, work in your local
   branch and push the changes to your fork. They will automatically appear in

>    the pull request.

Reviewer 3: I'd think we should also add a note here about how to rebase your repository, so you don't have to keep deleting your fork and reforking to make a new PR, while still being updated on what's new

Author: You mean by mentioning the following?

<img width="839" alt="image" src="https://user-images.githubusercontent.com/30755778/157272258-22430938-36bc-4c29-bd2c-1d6796593fdd.png">

Reviewer 3: Yep!

Author: Addressed in [2f39a31](https://github.com/huggingface/huggingface_hub/pull/692/commits/2f39a31f861f7bc081118e6ba14a3a36ed1abf98)

### Checklist

1. The title of your pull request should be a summary of its contribution;
2. If your pull request addresses an issue, please mention the issue number in
   the pull request description to make sure they are linked (and people
   consulting the issue know you are working on it);

> 3. To indicate a work in progress please prefix the title with `[WIP]`. These

Reviewer 3: Perhaps also state you should hit the "Draft PR" setting when submitting too? (Or are we saying we'll be the ones to go and do this)

5. Add high-coverage tests. No quality testing = no merge.
7. Due to the rapidly growing repository, it is important to make sure that no files that would significantly weigh down the repository are added. This includes images, videos and other non-text files. We prefer to leverage a hf.co hosted `dataset` like
   the ones hosted on [`hf-internal-testing`](https://huggingface.co/hf-internal-testing) in which to place these files and reference
   them by URL. We recommend putting them in the following dataset: [huggingface/documentation-images](https://huggingface.co/datasets/huggingface/documentation-images).
   If an external contribution, feel free to add the images to your PR and ask a Hugging Face member to migrate your images
   to this dataset.

### Tests

An extensive test suite is included to test the library behavior and several examples. Library tests can be found in
the [tests folder](https://github.com/huggingface/huggingface_hub/tree/main/tests).

The `huggingface_hub` library's normal behavior is to work with the production Hugging Face Hub. However,
for tests, we prefer to run on a staging version. In order to do this, it's important to set the 

> `HUGGINGFACE_CO_STAGING` environment variable to `1`.

Reviewer 3: ```suggestion
`HUGGINGFACE_CO_STAGING` environment variable to `1` when running tests. It is preferred to pass this in when running the tests, than setting a permanent environmental variable, as shown below.
```

WDYT

> We use `pytest` in order to run the tests. From the root of the
> repository, here's how to run tests with `pytest` for the library:
> 
> ```bash
> $ HUGGINGFACE_CO_STAGING=1 python -m pytest -sv ./tests
> ```

Reviewer 3: ```suggestion
We use `pytest` in order to run the tests for the library . From the root of the
repository they can be run with the following:

```bash
$ HUGGINGFACE_CO_STAGING=1 python -m pytest -sv ./tests
```
```

In fact, that's how `make test` is implemented (sans the `pip install` line)!

> You can specify a smaller set of tests in order to test only the feature

Reviewer 3: Perhaps a quick example of this? Or stating you should use `-k` to do so, etc.
